### DISCUSSION TITLE
Message aborted after 4 attempt(s)

### DISCUSSION THREAD

**Discussion by virtualuk** [tagyoureit/nodejs-poolController#432] [App and O/S Config Questions]
This is my first time getting nodejs-poolController (and nodejs-poolcontroller-dashboard) installed and up and running.  I'm running:

- RPi 3+ with latest desktop OS
- JBTek USB to RS485 adapter
- Pentair Sun Touch controller
- Pentair Intelliflow VS-3050 pump

I have the RS485 adapter connected to the Sun Touch and I was able to confirm that I was getting ffa5ff octets successfully.

When I fire up njspc it looks promising when it says "Pentair IntelliCom System Detected!" but then goes on to say "Found Controller Board undefined".  After that I get the following messages, with the aborted message being repeated 4-5 seconds:

`[5/2/2022, 6:23:48 PM] info: Requesting intellicom configuration
[5/2/2022, 6:23:52 PM] warn: Message aborted after 4 attempt(s): 165,1,16,33,197,1,0,1,157 
RS485 Stats:{ "bytesReceived": 121909 "success": 3, "failed": 147, "bytesSent": 72, "collisions": 0, "failureRate": 98.00% }
[5/2/2022, 6:23:56 PM] warn: Message aborted after 4 attempt(s): 165,1,16,33,202,1,0,1,162 
[5/2/2022, 6:24:01 PM] warn: Message aborted after 4 attempt(s): 165,1,16,33,202,1,1,1,163 
[5/2/2022, 6:24:05 PM] warn: Message aborted after 4 attempt(s): 165,1,16,33,202,1,2,1,164 
[5/2/2022, 6:24:09 PM] warn: Message aborted after 4 attempt(s): 165,1,16,33,202,1,3,1,165`

Can someone cast some light on what I'm doing wrong?

**Comment by rstrouse** [comment]
You are getting that message because we have not created a board signature for the SunTouch.  The SolarTouch turned out to not have any controllable features but it sounds like the SunTouch actually sends a status packet.  Could I get you to produce a replay that includes turning on and off the aux circuit?

https://github.com/tagyoureit/nodejs-poolController/wiki/How-to-capture-all-packets-for-issue-resolution

**Reply by ssorkin** [reply]
I'm in roughly the same boat. I have a SunTouch with an Intelliflow VS and and Intellichem. The requests from nodejs-poolController to the Suntouch are completely fruitless, but I can make things work by just listening to the chatter from the pump and Intellichem.

Would you like a replay from my system? I'd be happy to write a board signature for the SunTouch if you provide minimal guidance.

**Reply by rstrouse** [reply]
Sure it will shed some light on what is there and what isn't.

**Reply by ssorkin** [reply]
Here's my packet capture. 

Neither the pumps nor chemistry controller show up. As I mentioned, I can make them show up by doing the following:

1. IntelliChem: in IntelliChemStateMessage.ts, change line 25, getItemByAddress to create the item by adding "true, {isActive: true}" to the function call.
2. IntelliFlow: in PumpStateMessage.ts, change line 31, getPumpByAddress to do the same as above.

The code clearly wants the OCP to inform what components are present, but the SunTouch doesn't seem to have a desire to do so, which isn't surprising. I can't really abandon the SunTouch, since it provides the mechanism to read the temperatures and control the valves. I understand that Nixie can control the pump and read from the IntelliChem, but can it use the SunTouch for pure hardware control?

[2022-07-11_19-28-18.zip](https://github.com/tagyoureit/nodejs-poolController/files/9088974/2022-07-11_19-28-18.zip)

**Reply by rstrouse** [reply]
Can you please list all the equipment that SunTouch supports including valves and relays.  It is reporting some schedules.

**Reply by ssorkin** [reply]
Sure thing. Attached to mine is:

1. IntelliFlow (via RS485).
2. Booster Pump (via relay).
3. Solar valve (by the board itself).
4. Three temperature sensors (pool, air, solar).
5. IntelliChem (by RS485, no control, but the IntelliChem isn't chatty unless the SunTouch is on the bus; the SunTouch does display the ORP and pH in its diagnostics menu).

Additionally, it supports:

1. QT4 wireless remote.
2. IntelliChlor
3. 2 more valve actuators
4. pump relay, second aux relay, two speed pump relay

**Reply by rstrouse** [reply]
Is there any other way to access this system other than through the menus on the pad or the QT4 remote?

**Reply by rstrouse** [reply]
> The code clearly wants the OCP to inform what components are present, but the SunTouch doesn't seem to have a desire to do so, which isn't surprising. I can't really abandon the SunTouch, since it provides the mechanism to read the temperatures and control the valves. I understand that Nixie can control the pump and read from the IntelliChem, but can it use the SunTouch for pure hardware control?
> 

Yes Nixie can control your valves, circuits, pumps, chlorinator, IntelliChem, heater, lights, temperature probes ...etc.  There is nothing that SunTouch can do that Nixie can't do better.  It can use the existing high power relays.  I am pouring through the RS485 dump and do not see any way to set a circuit state (turn one on or off).  But I do see all the data related to all the equipment to include the QT4 configuration.  Yes you are right it appears that the supported equipment list is small enough that it probes for the items.

**Reply by ssorkin** [reply]
I only control it through the OCP itself. I have been able to toggle the circuits through nodejs poolcontroller. I believe that the abandoned Autelis device could control the SunTouch circuits.

**Reply by rstrouse** [reply]
Excellent.  I will put a board definition together using the Touch command protocols.  There does not seem to be any configuration abilities but we will see which functions work after we stand up the board,

**Reply by ssorkin** [reply]
I'm happy to test any of this, and apply any changes to the code live to my system. I found this other project (that I haven't tried yet), that claims to support adjusting thermostats and toggling circuits on the SunTouch (https://github.com/cvonk/OPNpool). Your code does seem to twiddle the thermostats and toggle the circuits, so it's unclear if there's anything new there.

**Reply by rstrouse** [reply]
There isn’t.  This project will also control the IntelliChem for you.

**Reply by tagyoureit** [reply]
That OPNpool has documented what they discovered on the RS485 (and as a bonus they put my name down as a contributor).  https://coertvonk.com/sw/embedded/opnpool-design/protocol-31965

I just skimmed it quickly but it's all very familiar.  It shouldn't be hard to get this working.  I bet @rstrouse is close to being done with it!

**Reply by rstrouse** [reply]
It looks like they remapped all of the identifiers so I need some information regarding your configuration.
I need some information about what you have things mapped to.
* For the circuits I need to know what circuit function and the name you have selected for Aux1-3
* For the valves what circuit do you have assigned to to Valve C.  It appears that valve A&B are assigned to intake and return respectively.
* For features I think they hard code these to Feature 1-4 and do not allow you to assign a circuit function (like Spillway).

**Reply by ssorkin** [reply]
The mappings are as follows:

1. Aux1 is set to the pressure side cleaner booster pump; nothing mapped to 2 and 3.
2. Valve C is mapped to solar; I don't use A and B.
3. I have no features in my system.

**Reply by tagyoureit** [reply]
@ssorkin is there any setting to tell SunTouch if this is a single body pool or shared body?

**Reply by rstrouse** [reply]
@tagyoureit there are 3 modes that it can operate in.  Single body, Solar Only, and Pool/Spa.  I suspect that this byte is on the Valve message and the Single body setting is factory set and it operates just like IntelliTouch where it simply changes the Pool and Spa circuits to Hi-Temp and Lo-Temp.

@ssorkin I need to know what the text says on the OCP for Aux 1.  Thank you for the valve clarification.  Related to the @tagyoureit question which mode are you operating in...  SOLAR ONLY or POOL/SPA?

Also I see one schedule which circuit is it supposed to be mapped to?

**Reply by ssorkin** [reply]
The OCP calls Aux 1 "AUX1" and this is also mapped to the down arrow key on the OCP.

I believe that I am operating in POOL/SPA mode.

**Reply by rstrouse** [reply]
I need to know what the text is for the circuit function on AUX 1-3
It will be one of the following values Generic, Cleaner, Spillway, Floor Cleaner, Heater, Solar.  Regardless of whether you have something wired to it I need to know what it is.

**Reply by ssorkin** [reply]
I can verify precisely when I get back home from vacation, but as I recall, there is no text for those circuits; they are displayed on the screen just as AUX_n_

**Reply by ssorkin** [reply]
This seems to be consistent with what the manual shows: https://www.pentair.com/content/dam/extranet/nam/pentair-pool/residential/manuals/automation/suntouch/suntouch-control-system-manual-english.pdf, (page 11 of the document, 19 of the PDF).

**Reply by rstrouse** [reply]
I know that SunTouch does not allow you to assign a name to the circuit.   However, there is an identifier that determines what the circuit does and I need this value specifically for AUX2 even if you do not have anything wired to it.   This is assigned to something that does not appear in EasyTouch or IntelliTouch.
![image](https://user-images.githubusercontent.com/47839015/178839725-68b30d38-2837-4d96-9087-1efc3c2cb7d2.png)

**Reply by ssorkin** [reply]
Ah, in this case, AUX1 on mine is set to “CLEANER” and no other ones are set.

**Reply by rstrouse** [reply]
Are you absolutely certain that AUX1 is set to that circuit function?  AUX1 and AUX3 appear to be set to GENERIC and AUX2 is set to 63 which when mapped points to a function type that does not exist on IntelliTouch or EasyTouch.  It might make sense if this was AUX2.  The other thing that they may have done is shifted the bits by 1 so AUX1 starts in the AUX2 position.  It appears as though SunTouch will respond that it has up to 10 circuits and features but supports only a combination of 7 of them.

**Reply by rstrouse** [reply]
You also have a schedule that runs from 12:48pm to 9:20pm on Sunday through Friday.   What circuit is supposed to run during that time?

**Reply by ssorkin** [reply]
1. I'll have to check about AUX1 when I return. It's certainly possible that AUX1 is GENERIC and AUX2 is set to cleaner, and that I never reset them when I made AUX1 control the cleaner relay. I'll let you know on Sunday.
2. I don't recall the schedules allowing day of week selection; I should only have one schedule set, to enable POOL (LOW) from 8 AM to 4 PM or similar.

**Reply by ssorkin** [reply]
I have verified the following on the OCP:

1. It is set to POOL/SPA mode.
2. Under circuit functions, AUX1 is GENERIC, AUX2 is SOLAR, AUX3 is GENERIC
3. Under valve functions, the only configurable valve is set to SOLAR.
4. Under schedule, I only have one schedule set: LOW from 8:00 AM to 3:00 PM, every day (there is no option to set days).

Let me know if you need any more information; I can twiddle the settings and capture more or test any code in situ.

**Reply by ssorkin** [reply]
Also, under the QuickTouch menu, the buttons are mapped to HIGH, AUX1, AUX2, AUX3.

**Reply by rstrouse** [reply]
Ok so pull njsPC and run another replay.  Please generate this replay with the Reload Config option checked.  This will let me know if there are any other items that are disabled between EasyTouch and SunTouch.

**Reply by ssorkin** [reply]
Here you go! One observation is that dashpanel shows that the spa is on, when the pool is currently on (in HIGH mode).
[replay.zip](https://github.com/tagyoureit/nodejs-poolController/files/9136483/replay.zip)

**Reply by rstrouse** [reply]
Alright so now you have me confused about what mode you are in.  Typically Pentair single body systems refer to two modes.  There is a HI-TEMP mode (HIGH) and a LO-TEMP (LOW) mode but this is only for single body systems.  The firmware is the same between these two systems except that they label the HI-TEMP circuit as the Spa and LO-TEMP as the Pool in shared body systems.  Are you certain that this is a shared Pool/Spa system?  Interestingly the status returns reflect this.

See page 40 of this manual.
https://www.pentair.com/content/dam/extranet/nam/pentair-pool/residential/manuals/automation/suntouch/suntouch-control-system-manual-english.pdf

Now that I have a replay that contains the entire config can I get you to perform a replay while doing the following from the SunTouch control panel.   It is important that you do this in order so I can be sure of the bit positions.  Do NOT check the checkbox for Capture Configuration Reload for this capture.

1. Start with all circuits and equipment off, then start the replay
2. Turn on the Spa, wait a couple of seconds, then turn it off
3. Turn on the Pool, wait a couple of seconds, then turn it off
4. Turn on Aux 1, wait a couple of seconds, the turn it off
5. Turn on Aux 2, wait a couple of seconds, then turn it off
6. Turn on Feature 1, wait a couple of seconds, then turn it off.
7. With everything still off change the system mode to Solar mode then back out of the advanced menu (see p7 of manual)
8. Change the system mode to Pool/Spa mode then back out of the advance menu.
9. End the replay and upload the results here.

**Reply by ssorkin** [reply]
Before I do that, let me note that you're almost certainly right. I checked the configuration and it's set to "pool/spa" (since the only other choice is "solar only" which takes away essentially all the configuration.

It's probably a bit of a bastard configuration since I don't have a spa, just a single body pool.

Do you still want me to go through this data collection?

**Reply by rstrouse** [reply]
Do not do steps 7 & 8.  I think you have a P/N 530820

Here is what I have been able to dig up re the SunTouch controllers
520857 - Booster Pump Version for single body and includes a booster pump relay
520856 - Solar system for single body and includes 2 temp sensors + 1 actuator
520859 - Single body system with 2 temp sensors
520819 - Single body system with 1 actuator, solar valve, and 2 temp sensors
520820 - Shared body pool/spa system with 2 actuators, 3 relays, and two temp sensors

Not sure what the differences in the firmware is but I suspect that it simply renames some of the menus.

**Reply by ssorkin** [reply]
I checked my invoice (from 2014, when this hardware was less than $300) and it indicates that I have a 520857.

**Reply by rstrouse** [reply]
Ok so I found a definitive way to determine which system you have.

Does your control panel look like this with a SPA button
![image](https://user-images.githubusercontent.com/47839015/179657586-1bcab35b-7c61-44d7-af2f-ba87ab900095.png)

Or does it look like this with a HI-TEMP button
![image](https://user-images.githubusercontent.com/47839015/179657906-b214b089-7e82-4797-b136-4d4a60704f16.png)

**Reply by ssorkin** [reply]
It says "HI-TEMP".

**Reply by rstrouse** [reply]
Well that settles it you have what SunTouch might consider a single body system without intake/return valves.

**Reply by ssorkin** [reply]
Do you still want me to capture steps 1-6 (or 1-8)?

**Reply by rstrouse** [reply]
1-6.  7 & 8 are no longer needed.   Now that I understand what I am looking at it makes more sense.

**Reply by ssorkin** [reply]
Here's the capture. Note: 1. all circuits except spa were turned on and off using the circuit on/off menu from the OCP, 2. I turned on "spa" (actually "high" here) using the up arrow on the OCP.

[replay 7-19-22.zip](https://github.com/tagyoureit/nodejs-poolController/files/9141877/replay.7-19-22.zip)

**Reply by rstrouse** [reply]
Pull njsPC I think I have all the state data reflecting appropriately.  I also added some configuration queries to the panel so I need to see the results there before I add the code to recognize the pump.

Can I get you to turn on and off every feature, aux circuit, and body from the OCP and verify that the status is correct in njsPC.  If it is all correct can I get you to try to turn on and off one of the circuits from dashPanel to see if SunTouch responds.  At this point I don't know if it will.

**Reply by ssorkin** [reply]
I have done so and everything works perfectly in both directions: the OCP state is reflected in dashPanel, and dashPanel can control every circuit. The only discrepancy I see now is the heat mode set on the OCP is not reflected on njsPC/dashPanel. specifically, solar is enabled but not reflected. I didn't dig into EquipmentStateMessage.ts, but it looks like it's equipped to read this, which is in byte 22 as shown in your code.

**Reply by rstrouse** [reply]
Do you know how to use messageManager?  If you watch the message returns when you change it maybe you can see the pattern.  It is possible that there is some strange packing on the heatMode.

**Reply by ssorkin** [reply]
I couldn't get messageManager to show any messages, but I realized that your code is working perfectly to read the heat settings. I had forgotten that I had turned off solar once my pool reached the low 90s. At this point the only problem is that the schedule isn't reflected correctly.

**Reply by rstrouse** [reply]
Here is a write up on messageManager.  https://github.com/rstrouse/nodejs-poolController-dashPanel/wiki/Message-Manager

I will take a closer look at the schedules.  Clearly there is some byte packing going on there.

**Reply by ssorkin** [reply]
Okay, I didn't realize that I needed to enable logging in the config menu to get rs485 messages to show up. It's working now and is simply amazing. Very nice work!

**Reply by rstrouse** [reply]
You can uncheck the log to file and console.  The filters only apply to the other selections.

**Reply by rstrouse** [reply]
Please pull njsPC and dashPanel.  Then can I get you to capture the configuration and upload that here.

I also am having a bit of trouble getting the schedules decoded.  Perhaps if you define another schedule for AUX1 from 8:00AM to 8:15AM (yes 15 min) I might be able to make sense of the data.  Please do this while capturing the replay.

**Reply by ssorkin** [reply]
I defined a schedule for AUX1 from 8:00AM to 8:15AM.

[replay 7-20-22.zip](https://github.com/tagyoureit/nodejs-poolController/files/9154482/replay.7-20-22.zip)

**Reply by rstrouse** [reply]
Can I get you to generate the replay with Capture Configuration Reload checked?

**Reply by ssorkin** [reply]
Here's a capture with the box checked. Note that the new schedule doesn't seem to show up in dashPanel.

[replay 7-20-22 with reload.zip](https://github.com/tagyoureit/nodejs-poolController/files/9155051/replay.7-20-22.with.reload.zip)

**Reply by ssorkin** [reply]
Where is the code to load the schedule? There seems to be a working protocol for SunTouch described here: https://coertvonk.com/sw/embedded/opnpool-design/protocol-31965 in the "Request Schedule" section.

**Reply by ssorkin** [reply]
There seems to be a working protocol for SunTouch described here: https://coertvonk.com/sw/embedded/opnpool-design/protocol-31965 in the "Request Schedule" section.

I see that you're requesting at 209 (0xd1) but SunTouch wants the request at 0xde.

**Reply by rstrouse** [reply]
Interesting that is typically the high speed circuits request.  But it does appear to have the output for at least two of the schedules.   The return 30 contains 2 schedules but there is not enough room for all 6 of them.
![image](https://user-images.githubusercontent.com/47839015/180105463-fd9dae52-0d5d-41dd-bcbd-7723dc5fb673.png)

Interestingly this is the return from the 222 request.
![image](https://user-images.githubusercontent.com/47839015/180105574-c3d1f0d2-22e8-4268-a3e0-a643357b1718.png)

**Reply by rstrouse** [reply]
Pull njsPC I added a something to SunTouchBoard.ts Line:234 to request 8 of these items.  Maybe there will be a totality of them with the extra returns.  If we manage to extract all of them we will then need to change `controller/comms/config/ScheduleMessage,ts` to process the action 30 message for the schedules.

Too bad I haven't seen anything for the pump configuration.  I may need to simply add some code to send out requests in the configuration range with all items we are not getting a return on to see if we can coerce it.

**Reply by ssorkin** [reply]
I pulled the code and reran and captured. See attached.

[replay 7-20-22 with reload 2.zip](https://github.com/tagyoureit/nodejs-poolController/files/9155348/replay.7-20-22.with.reload.2.zip)

**Reply by ssorkin** [reply]
I can see both of the schedules in reply to the 0xDE:

{"port":0,"id":243,"valid":true,"dir":"in","proto":"broadcast","for":[242],"pkt":[[255,255,255,255,255,255,255],[255,0,255],[165,1,15,16,30,16],[0,0,0,0,2,0,1,224,1,239,6,1,1,224,3,132],[4,53]],"ts": "2022-07-20T17:59:59.524-0700"}

**Reply by rstrouse** [reply]
Yes but there are a total of 6 potential.

**Reply by ssorkin** [reply]
Would you like me to add 4 more to see where they show up?

**Reply by ssorkin** [reply]
I have 6 total schedules in this dump.

In addition to LOW and AUX1, we have:

HIGH 1-2am
FT1 2-3am
FT2 3-4am
FT3 4-5am
[replay 7-20-22 6 schedules.zip](https://github.com/tagyoureit/nodejs-poolController/files/9155632/replay.7-20-22.6.schedules.zip)

**Reply by rstrouse** [reply]
So I fixed njsPC so that it will properly interpret the first 2 schedules.   However, we are missing some other items that should be appearing on the configuration.  I have added some exploratory requests to see if I can get the remaining items that we are missing from your configuration.

1. Schedules:  We only have returns for 2 of them.  There are 4 remaining somewhere but they are in none of the existing returns.
2. Pumps:  The pump definition is not being returned from the pump configuration requests
3. Chlorinator:  While your chlorinator is responding, its initial config is not being returned from SunTouch,

Given that the returns for the first 2 schedules are on a different request, I have added a series of configuration requests in the hopes that SunTouch responds to one or more of them with the missing configuration data.  Please pull njsPC and generate another replay with the Capture Configuration Reload checked.   I hope this blanket approach will yield the data we are looking for,

**Reply by ssorkin** [reply]
Here's the replay. The following schedules are set at the OCP:

1. LOW: 8am-3pm
2. FT1: 1-1:15am
3. FT2: 2-2:15am
4. FT3: 3-3:15am
5. FT4: 4-4:15am
6. AUX2: 5-5:15am

[replay 7-21-22.zip](https://github.com/tagyoureit/nodejs-poolController/files/9161882/replay.7-21-22.zip)

**Reply by rstrouse** [reply]
Ok so the pump configuration and additional schedules were not included with the last set of message probes.  I added a bunch more probes so if I can get you to pull njsPC again then capture another replay with the configuration reload checked.

Interestingly, I did find the request for the chlorinator but then realized you don't have one.  You are using IntelliChem and that is configuring just fine.  We are still missing pumps and 2/3 of the schedules.

**Reply by ssorkin** [reply]
Here you go.
[replay 7-21-22 v2.zip](https://github.com/tagyoureit/nodejs-poolController/files/9163379/replay.7-21-22.v2.zip)

**Reply by rstrouse** [reply]
Well I don't know where to go next.  I have probed every message within the normal range for requesting panel configuration.  The 4 of 6 schedules do not show up and neither does the pump configuration. I can add the pump if I see its status being reported but it will not allow you to change the speeds on the fly.  This will take some coding to disable all the functionality when we see a SunTouch.  

In the end you will get functionality to control the SunTouch but configuration will be view only.  Still that is a damn sight better than running out to the pad to turn on a manual circuit.  If it were me I would add a relay board to the pi and disable the SunTouch.  You will then have the ability to schedule items for the day of the week, sunset, sunrise.   Control certain functions of IntelliChem remotely.  Configure your pump remotely and have an unlimited number of potential circuits, feature, and groups.  It will also control IntelliBrite natively.

I think what they did to create the SunTouch is take an IntelliTouch motherboard and remove the personality board header.  Then the cut the code out of the firmware that allows configuration from an indoor panel.  The stuff that was left over were the things that

**Reply by ssorkin** [reply]
I'd love to be able to just see the pump in read only mode; I rarely, if ever change the speeds. Mostly I want to see the RPM and Watts, and hopefully the implied GPM (it's a simple formula for this pump) to know if I need to change the filter media.

**Reply by rstrouse** [reply]
If you pull njsPC.  It will now detect the pump on the first request for it to run.

**Comment by virtualuk** [comment]
Sure thing, will get one tomorrow when when I have some daylight.  Dumb
question in the interim, would it be possible for me to go direct to the
pump rather than the controller? My intent was to replace the controller
with njspc, but accessing the controller's RS-485 connection was the
easiest to access/identify to get things going.

**Reply by rstrouse** [reply]
Yes and replacing it will be a huge increase in capability.  If you keep it I was going to suggest that you use an aux njsPC port for the pump and let Nixie control it.  From what I see in the docs for the controller Nixie will offer more control options.

**Reply by rstrouse** [reply]
@virtualuk if you haven't already replaced your SunTouch can I get a replay?
(https://github.com/tagyoureit/nodejs-poolController/wiki/How-to-capture-all-packets-for-issue-resolution)

And some information about your controller.
* Which mode is set on the panel Pool/Spa or Solar Only?
* In the OCP configuration which circuit function is assigned to each AUX1-3 [ GENERIC,
CLEANER, SPILLWAY, FLOOR CLEANER, HEATER, SOLAR]?
* How many features do you have defined?
* Describe each Schedule you have set up in the OCP [Circuit, Start Time, Stop Time].

**Reply by rstrouse** [reply]
@virtualuk are you still around.  I have taken another user's replay and created a board definition.  It would be great if your panel is actually a shared body panel.
